src/crimson/os/seastore/seastore_types.h

```diff
@@ -37,26 +37,121 @@ struct seastore_meta_t {
+// mask for segment manager id
+constexpr segment_id_t SM_ID_MASK =
+  0xF << (std::numeric_limits<segment_id_t>::digits - device_id_len);
+
+using device_segment_id_t = segment_id_t;
+
+constexpr device_segment_id_t MAX_SEG_ID =
+  std::numeric_limits<device_segment_id_t>::max();
+constexpr device_segment_id_t NULL_SEG_ID =
+  std::numeric_limits<device_segment_id_t>::max() - 1;
 /* Used to denote relative paddr_t */
-constexpr segment_id_t RECORD_REL_SEG_ID =
-  std::numeric_limits<segment_id_t>::max() - 2;
-constexpr segment_id_t BLOCK_REL_SEG_ID =
-  std::numeric_limits<segment_id_t>::max() - 3;
+constexpr device_segment_id_t RECORD_REL_SEG_ID =
+  std::numeric_limits<device_segment_id_t>::max() - 2;
+constexpr device_segment_id_t BLOCK_REL_SEG_ID =
+  std::numeric_limits<device_segment_id_t>::max() - 3;
 // for tests which generate fake paddrs
-constexpr segment_id_t FAKE_SEG_ID =
-  std::numeric_limits<segment_id_t>::max() - 4;
-
+constexpr device_segment_id_t FAKE_SEG_ID =
+  std::numeric_limits<device_segment_id_t>::max() - 4;
 /* Used to denote references to notional zero filled segment, mainly
  * in denoting reserved laddr ranges for unallocated object data.
  */
-constexpr segment_id_t ZERO_SEG_ID =
+constexpr device_segment_id_t ZERO_SEG_ID =
   std::numeric_limits<segment_id_t>::max() - 5;
+constexpr device_segment_id_t DELAYED_TEMP_SEG_ID =
+  std::numeric_limits<segment_id_t>::max() - 6;
+
+struct device_segment_t {
+  device_segment_id_t segment;
+
+  device_segment_t() = default;
+  constexpr device_segment_t(device_segment_id_t segment)
+    : segment(segment) {}
+  device_segment_t(device_id_t id, segment_id_t segment);
+
+  [[gnu::always_inline]]
+  device_id_t device_id() const {
+    return get_device_id(segment);
+  }
+
+  [[gnu::always_inline]]
+  segment_id_t segment_id() const {
+    return strip_device_id(segment);
+  }
+
+  bool operator==(const device_segment_t& other) const {
+    return segment == other.segment;
+  }
+  bool operator!=(const device_segment_t& other) const {
+    return segment != other.segment;
+  }
+  bool operator<(const device_segment_t& other) const {
+    return segment < other.segment;
+  }
+  bool operator<=(const device_segment_t& other) const {
+    return segment <= other.segment;
+  }
+  bool operator>(const device_segment_t& other) const {
+    return segment > other.segment;
+  }
+  bool operator>=(const device_segment_t& other) const {
+    return segment >= other.segment;
+  }
+  device_segment_t& operator++() {
+    ++segment;
+    return *this;
+  }
+  device_segment_t operator++(int) {
+    device_segment_t old = *this;
+    ++segment;
+    return old;
+  }
+  device_segment_t& operator--() {
+    --segment;
+    return *this;
+  }
+  device_segment_t operator--(int) {
+    device_segment_t old = *this;
+    --segment;
+    return old;
+  }
+  device_segment_t operator+(device_segment_id_t t) {
+    return {segment + t};
+  }
+
+  DENC(device_segment_t, v, p) {
```

Comment:
We don't need compatibility here, probably easiest to write the encode/decode/bound_encode directly and skip the DENC macro (see the DENC macro definition in common/denc.h).